src/core/inductive.ml

```diff
@@ -0,0 +1,213 @@
+(** Generating of induction principles. *)
+
+open Timed
+open Pos
+open Console
+open Terms
+open Print
+open Syntax
+
+(** Logging function for generating of inductive principle. *)
+let log_ind = new_logger 'g' "indu" "generating induction principle"
+let log_ind = log_ind.logger
+
+(** Builtin configuration for induction. *)
+type config =
+  { symb_Prop : sym (** : TYPE. Type of propositions. *)
+  ; symb_prf  : sym (** : Prop → TYPE.
+                        Interpretation of propositions as types. *) }
+
+(** [get_config ss pos] build the configuration using [ss]. *)
+let get_config : Sig_state.t -> Pos.popt -> config = fun ss pos ->
+  let builtin = Builtin.get ss pos in
+  { symb_Prop = builtin "Prop"
+  ; symb_prf  = builtin "P" }
+
+(** [fold_cons_typ pos codom get_name build_rec_hyp domrec dom ind_sym
+    cons_sym f_rec f_not_rec acc] generates some value iteratively by going
+    through the structure of [cons_sym.sym_type]. The position [pos] is the
+    position of the command inductive where the constructors were defined.
+    On the codomain, the function [codom] is called on the arguments to which
+    [ind_sym] is applied and the variables of the products (in reverse order).
+    In case of a product, the function [get_name] splits the product and
+    gives a name to the variable if it has none. In case of recursive
+    occurrences, the function [build_rec_hyp] builds the recursive hypothesis
+    associated, and then the function [domrec] is applied to conclude the
+    building. Otherwise, the function [dom] is called. If you would like to
+    store a temporary result, the initial value is [acc], and you can change
+    this value in the recursive case with the function [f_rec], and on the
+    other case with the function [f_not_rec]. *)
+let fold_cons_typ (pos : popt) (codom : term list -> 'b list -> 'a -> 'c)
+      (get_name : tbinder -> int -> 'b * term )
+      (build_rec_hyp : term list -> 'b -> 'a)
+      (domrec : term -> 'b -> 'a -> 'c -> 'c) (dom : term -> 'b -> 'c -> 'c)
+      (ind_sym : sym) (cons_sym : sym) (f_rec : 'a -> 'b -> 'a -> 'a)
+      (f_not_rec : 'a -> 'b -> 'a) (acc : 'a) : 'c =
+  let rec aux : 'b list -> 'a -> int -> term -> 'c = fun xs p i a ->
```

Comment:
We have the invariant that `i = List.length xs`. So this argument could be removed by passing xs as argument to get_name (or inj_var).

Reply:
Done